### ISSUE TITLE
In JS, assertions imported via commonjs have bogus error

### ISSUE THREAD

**Issue by sandersn** [microsoft/TypeScript#38379]
**Code**

```ts
// @filename: declares.d.ts
// this is the declaration for 'assert' from @types/node
export function art(value: any, message?: string | Error): asserts value;

// @filename: test.js
const { art } = require('./declares')
let x = 1
art(x)
```

**Expected behavior:**
No error.

**Actual behavior:**
Error: "Assertions require every name in the call target to be declared with an explicit type annotation."

There is no error when using ES imports:

```js
// @filename: test2.js
import { art } from './declares'
let x = 1
art(x)
```

**Comment by sandersn** [comment]
Repros since 3.7, when we introduced asserts.

**Comment by DanielRosenwasser** [comment]
My hot take is that this is a bogus error almost everywhere you could get it...

**Comment by Raynos** [comment]
I ran into this today when updating `@types/node` and using `require('assert')` from node core.

I fixed this with

```js
/** @type {import('assert')} */
const assert = require('assert')
```

Is it possible for the type definition of `require` in TypeScript itself to return `function require(module: T): import<T>` ?

**Comment by sandersn** [comment]
The problem is that symbols imported via `require` aren't actually aliases, so they don't get resolved in `getTypeOfDottedName`.

**Comment by sandersn** [comment]
I prototyped a solution at the branch alias-for-require. It fixes this example, but is woefully incomplete everywhere else. It still needs:

1. More complete handling of object literal syntax, which has way more options than the import syntax.
2. Handling of un-destructured require.
2. Rewrite of d.ts emit code.
3. Replacement and removal of all the special-case fake-alias code elsewhere.
5. Anything else I don't know about right now.

@weswigham you might be interested in this.
